File: test_basic_block.py
#!/usr/bin/env -S uv run
"""
Test: test_basic_block
Tests LLVM BasicBlock creation and manipulation via Python bindings.

This is the Python equivalent of tests/test_basic_block.cpp.
Output should match the C++ golden master test.

LLVM APIs covered (via Python bindings):
- append_basic_block()
- name property
- parent property
- entry_block, basic_block_count
- first_basic_block, next_block, last_basic_block
- first_instruction, last_instruction, terminator
- move_before(), move_after()
- create_basic_block() (unattached), append_existing_basic_block()
"""

import sys
import llvm


def main():
    with llvm.create_context() as ctx:
        with ctx.create_module("test_basic_block") as mod:
            i32 = ctx.int32_type()
            void_ty = ctx.void_type()

            # Create a function to hold basic blocks
            func_ty = ctx.function_type(void_ty, [], vararg=False)
            func = mod.add_function("test_func", func_ty)

            # Append basic blocks
            entry = func.append_basic_block("entry", ctx)
            middle = func.append_basic_block("middle", ctx)
            exit_bb = func.append_basic_block("exit", ctx)

            # Get block names
            entry_name = entry.name
            middle_name = middle.name
            exit_name = exit_bb.name

            # Get parent function
            entry_parent = entry.parent

            # Get entry basic block
            func_entry = func.entry_block
            assert func_entry is not None

            # Count basic blocks
            bb_count = func.basic_block_count

            # Create a builder to add instructions
            with ctx.create_builder() as builder:
                # Add instructions to entry block
                builder.position_at_end(entry)
                builder.br(middle)

                # Add instructions to middle block
                builder.position_at_end(middle)
                builder.br(exit_bb)

                # Add instructions to exit block
                builder.position_at_end(exit_bb)
                builder.ret_void()

                # Get first/last instructions
                entry_first = entry.first_instruction
                entry_last = entry.last_instruction
                exit_terminator = exit_bb.terminator

                # Create unattached block
                unattached = ctx.create_basic_block("unattached")

                # Attach unattached block to function
                func.append_existing_basic_block(unattached)
                builder.position_at_end(unattached)
                builder.unreachable()

            # Block count after adding unattached
            bb_count_after = func.basic_block_count

            # Move blocks around: move unattached before exit
            unattached.move_before(exit_bb)

            # Verify module
            if not mod.verify():
                print(
                    f"; Verification failed: {mod.get_verification_error()}",
                    file=sys.stderr,
                )
                return 1

            # Print diagnostic comments
            print("; Test: test_basic_block")
            print(";")
            print("; Basic block info:")
            print(f";   entry name: {entry_name}")
            print(f";   middle name: {middle_name}")
            print(f";   exit name: {exit_name}")
            print(";")
            print("; Parent checks:")
            # Compare by name since we can't compare object identity directly
            print(
                f";   entry parent is func: {'yes' if entry_parent.name == func.name else 'no'}"
            )
            print(
                f";   func entry block is entry: {'yes' if func_entry.name == entry.name else 'no'}"
            )
            print(";")
            print("; Block counts:")
            print(f";   initial count: {bb_count}")
            print(f";   after adding unattached: {bb_count_after}")
            print(";")
            print("; Instruction checks:")
            print(
                f";   entry has first instruction: {'yes' if entry_first is not None else 'no'}"
            )
            print(
                f";   entry first == last (single inst): {'yes' if entry_first is not None and entry_last is not None else 'no'}"
            )
            print(
                f";   exit has terminator: {'yes' if exit_terminator is not None else 'no'}"
            )
            print(";")
            print("; Block iteration (after move):")
            for i, bb in enumerate(func.basic_blocks):
                print(f";   [{i}] {bb.name}")

            # Get last block
            last_bb = func.last_basic_block
            assert last_bb is not None
            print(";")
            print(f"; Last block: {last_bb.name}")

            print()

            # Print module IR
            print(mod.to_string(), end="")

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: tests/test_basic_block.cpp
/**
 * Test: test_basic_block
 * Tests LLVM BasicBlock creation and manipulation
 *
 * LLVM-C APIs covered:
 * - LLVMAppendBasicBlockInContext()
 * - LLVMGetBasicBlockName()
 * - LLVMGetBasicBlockParent()
 * - LLVMGetEntryBasicBlock()
 * - LLVMCountBasicBlocks()
 * - LLVMGetFirstBasicBlock(), LLVMGetNextBasicBlock(), LLVMGetLastBasicBlock()
 * - LLVMGetFirstInstruction(), LLVMGetLastInstruction()
 * - LLVMGetBasicBlockTerminator()
 * - LLVMMoveBasicBlockBefore(), LLVMMoveBasicBlockAfter()
 * - LLVMCreateBasicBlockInContext()
 * - LLVMDeleteBasicBlock()
 */

#include <cstdio>
#include <llvm-c/Analysis.h>
#include <llvm-c/Core.h>

int main() {
  LLVMContextRef ctx = LLVMContextCreate();
  LLVMModuleRef mod =
      LLVMModuleCreateWithNameInContext("test_basic_block", ctx);

  LLVMTypeRef i32 = LLVMInt32TypeInContext(ctx);
  LLVMTypeRef void_ty = LLVMVoidTypeInContext(ctx);

  // Create a function to hold basic blocks
  LLVMTypeRef func_ty = LLVMFunctionType(void_ty, nullptr, 0, 0);
  LLVMValueRef func = LLVMAddFunction(mod, "test_func", func_ty);

  // Append basic blocks
  LLVMBasicBlockRef entry = LLVMAppendBasicBlockInContext(ctx, func, "entry");
  LLVMBasicBlockRef middle = LLVMAppendBasicBlockInContext(ctx, func, "middle");
  LLVMBasicBlockRef exit_bb = LLVMAppendBasicBlockInContext(ctx, func, "exit");

  // Get block names
  const char *entry_name = LLVMGetBasicBlockName(entry);
  const char *middle_name = LLVMGetBasicBlockName(middle);
  const char *exit_name = LLVMGetBasicBlockName(exit_bb);

  // Get parent function
  LLVMValueRef entry_parent = LLVMGetBasicBlockParent(entry);

  // Get entry basic block
  LLVMBasicBlockRef func_entry = LLVMGetEntryBasicBlock(func);

  // Count basic blocks
  unsigned bb_count = LLVMCountBasicBlocks(func);

  // Create a builder to add instructions
  LLVMBuilderRef builder = LLVMCreateBuilderInContext(ctx);

  // Add instructions to entry block
  LLVMPositionBuilderAtEnd(builder, entry);
  LLVMBuildBr(builder, middle);

  // Add instructions to middle block
  LLVMPositionBuilderAtEnd(builder, middle);
  LLVMBuildBr(builder, exit_bb);

  // Add instructions to exit block
  LLVMPositionBuilderAtEnd(builder, exit_bb);
  LLVMBuildRetVoid(builder);

  // Get first/last instructions
  LLVMValueRef entry_first = LLVMGetFirstInstruction(entry);
  LLVMValueRef entry_last = LLVMGetLastInstruction(entry);
  LLVMValueRef exit_terminator = LLVMGetBasicBlockTerminator(exit_bb);

  // Create unattached block
  LLVMBasicBlockRef unattached =
      LLVMCreateBasicBlockInContext(ctx, "unattached");

  // Attach unattached block to function
  LLVMAppendExistingBasicBlock(func, unattached);
  LLVMPositionBuilderAtEnd(builder, unattached);
  LLVMBuildUnreachable(builder);

  // Block count after adding unattached
  unsigned bb_count_after = LLVMCountBasicBlocks(func);

  // Move blocks around: move unattached before exit
  LLVMMoveBasicBlockBefore(unattached, exit_bb);

  // Iterate through blocks
  int block_index = 0;
  printf("; Test: test_basic_block\n");
  printf(";\n");
  printf("; Basic block info:\n");
  printf(";   entry name: %s\n", entry_name);
  printf(";   middle name: %s\n", middle_name);
  printf(";   exit name: %s\n", exit_name);
  printf(";\n");
  printf("; Parent checks:\n");
  printf(";   entry parent is func: %s\n", entry_parent == func ? "yes" : "no");
  printf(";   func entry block is entry: %s\n",
         func_entry == entry ? "yes" : "no");
  printf(";\n");
  printf("; Block counts:\n");
  printf(";   initial count: %u\n", bb_count);
  printf(";   after adding unattached: %u\n", bb_count_after);
  printf(";\n");
  printf("; Instruction checks:\n");
  printf(";   entry has first instruction: %s\n",
         entry_first != nullptr ? "yes" : "no");
  printf(";   entry first == last (single inst): %s\n",
         entry_first == entry_last ? "yes" : "no");
  printf(";   exit has terminator: %s\n",
         exit_terminator != nullptr ? "yes" : "no");
  printf(";\n");
  printf("; Block iteration (after move):\n");
  for (LLVMBasicBlockRef bb = LLVMGetFirstBasicBlock(func); bb;
       bb = LLVMGetNextBasicBlock(bb)) {
    printf(";   [%d] %s\n", block_index++, LLVMGetBasicBlockName(bb));
  }

  // Get last block
  LLVMBasicBlockRef last_bb = LLVMGetLastBasicBlock(func);
  printf(";\n");
  printf("; Last block: %s\n", LLVMGetBasicBlockName(last_bb));

  LLVMDisposeBuilder(builder);

  // Verify module
  char *error = nullptr;
  if (LLVMVerifyModule(mod, LLVMReturnStatusAction, &error)) {
    fprintf(stderr, "; Verification failed: %s\n", error);
    LLVMDisposeMessage(error);
    LLVMDisposeModule(mod);
    LLVMContextDispose(ctx);
    return 1;
  }
  LLVMDisposeMessage(error);

  printf("\n");

  // Print module IR
  char *ir = LLVMPrintModuleToString(mod);
  printf("%s", ir);
  LLVMDisposeMessage(ir);

  // Cleanup
  LLVMDisposeModule(mod);
  LLVMContextDispose(ctx);

  return 0;
}
